### PR TITLE
chore: use efficient persona id query path

### PR DESCRIPTION
**Pull request by evan-onyx**
## Description

Previously we passed user file ids down and did a giant OR (file_id = id1 OR file_id = id2 OR...) which doesn't scale well to large numbers of files. The updated path has each chunk know which persona(s) it is attached to, so the query is just AND persona_id = <persona_id> AND ... (although there are many cases, check out FILTER_SEMANTICS).

We waited a few weeks to enable this path so more people would have time to migrate to the new more efficient query pattern for user files attached to assistants. There will inevitably be people who do huge upgrades and temporarily aren't able to search for userfiles attached to assistants at some point in the future, but it's a relatively rare edge case (they need to oveflow the context window) that also gets solved by a metadata syncing task within a day or so.

## How Has This Been Tested?

CI has a bunch of tests that check this behavior

## Additional Options

- [ ] [Optional] Please cherry-pick this PR to the latest release version.
- [x] [Optional] Override Linear Check

---
## Summary by cubic
Switch knowledge filtering to a persona-based scope instead of OR-ing per-file IDs. `persona_id` now triggers the knowledge scope; `project_id` only widens an existing scope. This makes queries faster and simpler across OpenSearch and Vespa.

- **Refactors**
  - Removed `user_file_ids` from `IndexFilters`, pipeline, and query builders; `persona_id` starts the scope, `project_id` is additive-only.
  - Updated OpenSearch and Vespa filters to OR `document_set`, `attached_document_ids`/`hierarchy_node_ids` (OpenSearch), and `persona_id`; add `project_id` only when a scope exists.
  - Simplified federated retrieval by dropping the user-file short-circuit; connectors are no longer skipped based on user files alone.
  - Stopped passing file IDs in tools; updated tests to assert persona-scoped filtering; refreshed `FILTER_SEMANTICS.md` to document the new rules.

<sup>Written for commit a4c9926eb136a2222898d726179d2d2024ac636f. Summary will update on new commits.</sup>

### CODE REVIEW

File: backend/onyx/context/search/retrieval/search_runner.py [hunk 1]
```diff
@@ -110,7 +110,9 @@ def search_chunks(
             user_id=user_id,
             source_types=list(source_filters) if source_filters else None,
             document_set_names=query_request.filters.document_set,
-            user_file_ids=query_request.filters.user_file_ids,
+            has_user_files=bool(
+                query_request.filters.persona_id or query_request.filters.project_id
+            ),
```
[the remark this answers is not in the document]
Comment (evan-onyx) [reply backend/onyx/context/search/retrieval/search_runner.py]: good catch, but the solution is actually to remove this flag entirely. It was used to distinguish between use cases (user files vs everything else) but we now allow searching over both.

File: backend/onyx/context/search/retrieval/search_runner.py [hunk 2]
```diff
@@ -110,7 +110,9 @@ def search_chunks(
...
         )
```
[the remark this answers is not in the document]
Comment (evan-onyx) [reply backend/onyx/context/search/retrieval/search_runner.py]: good catch, but the solution is actually to remove this flag entirely. It was used to distinguish between use cases (user files vs everything else) but we now allow searching over both.

File: backend/onyx/federated_connectors/federated_retrieval.py [hunk 3]
```diff
@@ -38,15 +38,15 @@ def get_federated_retrieval_functions(
...
     document_set_names: list[str] | None,
     slack_context: SlackContext | None = None,
-    user_file_ids: list[UUID] | None = None,
+    has_user_files: bool = False,
 ) -> list[FederatedRetrievalInfo]:
     # When User Knowledge (user files) is the only knowledge source enabled,
     # skip federated connectors entirely. User Knowledge mode means the agent
     # should ONLY use uploaded files, not team connectors like Slack.
-    if user_file_ids and not document_set_names:
+    if has_user_files and not document_set_names:
```
[the remark this answers is not in the document]
Comment (evan-onyx) [reply backend/onyx/federated_connectors/federated_retrieval.py]: as above, we're just removing this and allowing both.

File: backend/onyx/context/search/models.py [hunk 4]
```diff
@@ -70,7 +69,6 @@ class BaseFilters(BaseModel):
 
 
 class UserFileFilters(BaseModel):
-    user_file_ids: list[UUID] | None = None
     project_id: int | None = None
     persona_id: int | None = None
```
[the remark this answers is not in the document]
Comment (evan-onyx) [reply backend/onyx/context/search/models.py:73]: nah these are still filters used for user files

File: backend/onyx/document_index/opensearch/search.py [hunk 5]
```diff
@@ -1074,24 +1054,20 @@ def _get_hierarchy_node_filter(
...
-            # only when an explicit restriction is already in effect.
-            if project_id is not None:
-                knowledge_filter["bool"]["should"].append(
-                    _get_user_project_filter(project_id)
-                )
             if persona_id is not None:
                 knowledge_filter["bool"]["should"].append(
                     _get_persona_filter(persona_id)
                 )
+            # Additive: widen scope to also cover overflowing project
```
Comment (acaprau) [inline backend/onyx/document_index/opensearch/search.py:1065]: idk why this was moved down, the code looks identical, not a problem just not sure if this was intentional or not.
